File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        res = []

        def getPermutations(current, used):

            if len(nums) == len(current):
                res.append(current.copy())
                return

            for i in range(len(nums)):
                if nums[i] not in used:
                    used.add(nums[i])
                    current.append(nums[i])
                    getPermutations(current, used)
                    current.pop()
                    used.remove(nums[i])

        getPermutations([], set())

        return res

File: 0046-permutations/0046-permutations.java
class Solution {

    public List<List<Integer>> permute(int[] nums) {

        List<List<Integer>> result = new ArrayList<List<Integer>>();
        Set<Integer> used = new HashSet<Integer>();
        List<Integer> current = new ArrayList<>();

        getPermutations(nums, current, used, result);

        return result;
    }

    public void getPermutations(int[] nums, List<Integer> current, Set<Integer> used, List<List<Integer>> result){

        if(nums.length == current.size()){
            result.add(new ArrayList<Integer>(current));
            return;
        }

        for(int i=0; i<nums.length; i++){
            if(used.add(nums[i])){ 
                current.add(nums[i]);
                getPermutations(nums, current, used, result);
                current.remove(current.size()-1);
                used.remove(nums[i]);
            }
        }
    }
}